### PR TITLE
minor: upgrade to apollo/client (v3)

### PR DESCRIPTION
**Pull request by sirugh**
Signed-off-by: sirugh <rugh@adobe.com>

## Description

Upgrades to @apollo/client v3.
Adds `zen-observable-ts` as a dependency, which should have been done earlier.
Updates `typescript` to v4.
Changed tsconfig as required by upgrades.

## Motivation and Context

See #5

## How Has This Been Tested?

1. Check out this PR.
2. Run `yarn build` and makes sure there is no errors.
3. `yarn link`
4. In pwa-studio, run `yarn link @adobe/apollo-link-mutation-queue`
5. In pwa-studio, run `yarn watch` and make sure the server builds/runs.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by sirugh** [comment]
@paales :)

File: package.json [hunk 1]
```diff
@@ -18,11 +18,15 @@
   "author": "Stephen Rugh <rugh@adobe.com>",
   "license": "Apache-2.0",
   "dependencies": {
-    "apollo-link": "~1.2.13"
+    "@apollo/client": "~3.3.15",
+    "react": "^17.0.2",
+    "react-refresh": "^0.10.0",
```
Comment (sirugh) [inline package.json:23]: Seems like apollo client expects you to have react and react-refresh. Without these dependencies (even as peerDeps) I could not get pwa studio to run.

File: src/MutationQueueLink.ts [hunk 2]
```diff
@@ -9,14 +9,8 @@ the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTA
...
-import {
-  ApolloLink,
-  Observable,
-  Operation,
-  FetchResult,
-  NextLink,
-} from "apollo-link";
-import { Observer } from "zen-observable-ts";
+import { ApolloLink, Operation, FetchResult, NextLink } from "@apollo/client";
+import { Observer, Observable } from "zen-observable-ts";
```
Comment (sirugh) [inline src/MutationQueueLink.ts:13]: Main change here was the upgrade to @apollo/client and moving `Observer` to this line, which is where it came from anyways in `apollo-link`.

File: tsconfig.json [hunk 3]
```diff
@@ -1,10 +1,11 @@
 {
   "compilerOptions": {
+    "allowSyntheticDefaultImports": true,
     "declaration": true,
     "module": "commonjs",
     "moduleResolution": "node",
     "noImplicitAny": true,
-    "lib": ["esnext"],
+    "lib": ["dom", "esnext"],
```
Comment (sirugh) [inline tsconfig.json:8]: This and the above change are required, otherwise ts pukes.

**Review by jimbo — APPROVED** [review]

**Pull request merged**